### ISSUE TITLE
Increasing a stake leads to lower reward for first mint after increase - problem resolves itself after that first period

### ISSUE THREAD

**Issue by derekpierre** [nucypher/nucypher#2691]
When an existing stake size is increased (eg. via `nucypher stake increase`), the first `mint` that occurs produces less staking rewards than expected.

The sequence of events looks like the following:
* Assume there is a sub-stake, 1) with duration of 52 periods
*Assume sub-take 1) has restake on, and winddown off
* Sub-stake 1) has X NU tokens in it
* An increase operation is performed to increase the stake size from X to Y, where Y > X.
* The increase operation will produce a temporary sub-stake, 2), with value X that has a start period for the current period, and ends/expires/unlocks after the current period. Sub-stake 1) has a value of Y, but a start date beginning in the next period, and an end date 52 periods from then (of course winddown is off, so it's a moving window).

Therefore, sub-stake 2) is applicable for the current period (based on start date), and sub-stake 1) is valid for the subsequent periods (based on start date). Therefore, the first `mint` that occurs after the increase will use sub-stake 2) for calculating the rewards that should be minted for the staker. However, because sub-stake 2) is applicable for this period (based on start date), but only has a 1-period duration, the kappa coefficient used for calculating rewards is understated, and the stake receives fewer rewards than it should have for that period.

Here's an example of such a stake on mainnet; look at the "Boost" value (I used code from https://github.com/nucypher/nucypher/pull/2690):
![Screen Shot 2021-05-11 at 2 58 32 PM](https://user-images.githubusercontent.com/19150641/117874210-b0f62480-b26e-11eb-8939-b4314235a6e3.png)
You can see the reduced Boost value for the temporary sub-stake (# 1 in the image). The boost of 1.02 in the image is because the duration is 1 period so `kappa = 0.5 * (1 + 1/52) = 0.5096`, and therefore `boost = 0.5096 / 0.5 = 1.02`.

*Side Note: Sub-stake 2) may also have effects on NU received as staking rewards being unlocked instead of being restaked even though restaking is on...this I am less certain of at this time - but it is something I've noticed with people who have done increases, and is only just a theory at this point.*

Anyway, so for subsequent periods, the rewards calculation resolves itself because sub-stake 1) becomes applicable (based on start date), has the increased value of staked NU (i.e. Y), and the correct duration. So the problem is only localized to the first period immediately after an increase.

I wrote a very rough test that shows such a failure - `test_reward_ratio_after_increase_first_phase` as part of https://github.com/derekpierre/nucypher/commit/522914f46d3848a9f4579144a5d56751e011d002 - in my `reward-ratio` branch.  The other tests added pass, but was just to prove that the ratio of rewards holds. One caveat is that tests must be run individually since they are all in the same module and `testerchain` is per module, and it wasn't clear how to make a new testerchain for each test. I tried creating a new one per test, but that failed and I didn't look too deeply into it - so run `pytest -s tests/contracts/integration/test_contract_economics.py::test_reward_ratio_after_increase_first_phase`

In this test, two stakers have the same stake size and stake settings. For some periods these two stakers receive the same staking rewards. At a random time staker 2 increases their stake to now have a stake 2x the size of the staker 1. Therefore, the first period after the increase is performed both stakers should receive the same staking rewards (last period that they both have the same stake size), and then for subsequent periods after that staker 2 should receive 2x the rewards that staker 1 receives.

The test will fail and shows that the first period after the increase the staker that increased their stake received less rewards even though they staked the same amount that period. However, note that after this first period staker 2 does indeed receive 2x the staking rewards of staker 1 (this can be seen by setting `skipAssertionsAfterIncrease` on L202 to True which will print reward values. Example of test output (with `skipAssertionsAfterIncrease` set to True):

```
>>>> Increase ursula2 NU in period 26
...
>>> ursula1 reward calculated during period 27 for 26: 3508783.178082191575906268
>>> ursula2 reward calculated during period 27 for 26: 1739391.589041095787953134
>>>> base rewards at increase 92967754.219178076761516102
...
>>> ursula1 reward calculated during period 28 for 27: 2339188.785388127717270845
>>> ursula2 reward calculated during period 28 for 27: 4678377.570776255434541691
>>> ursula1 rewards since increase: 4108580.374429223505223979
>>> ursula2 rewards since increase: 4693377.570776255434541691
...
>>> ursula1 reward calculated during period 29 for 28: 2339188.785388127717270845
>>> ursula2 reward calculated during period 29 for 28: 4678377.570776255434541691
>>> ursula1 rewards since increase: 6447769.159817351222494824
>>> ursula2 rewards since increase: 9371755.141552510869083382
...
>>> ursula1 reward calculated during period 30 for 29: 2339188.785388127717270845
>>> ursula2 reward calculated during period 30 for 29: 4678377.570776255434541691
>>> ursula1 rewards since increase: 8786957.945205478939765669
>>> ursula2 rewards since increase: 14050132.712328766303625073
...
>>> ursula1 reward calculated during period 31 for 30: 2339188.785388127717270845
>>> ursula2 reward calculated during period 31 for 30: 4678377.570776255434541691
>>> ursula1 rewards since increase: 11126146.730593606657036514
>>> ursula2 rewards since increase: 18728510.283105021738166764
...
>>> ursula1 reward calculated during period 32 for 31: 2339188.785388127717270845
>>> ursula2 reward calculated during period 32 for 31: 4678377.570776255434541691
>>> ursula1 rewards since increase: 13465335.515981734374307359
>>> ursula2 rewards since increase: 23406887.853881277172708455
```

As you can see the increase occurred in period 26. The minting for period 26 occurring during period 27 yielded the following:
```
>>> ursula1 reward calculated during period 27 for 26: 3508783.178082191575906268
>>> ursula2 reward calculated during period 27 for 26: 1739391.589041095787953134
```

So ursula2 received a lower reward than ursula1 for the first period after the increase.

For subsequent periods the following is repeated:
```
>>> ursula1 reward calculated during period x for x-1: 2339188.785388127717270845
>>> ursula2 reward calculated during period x for x-1: 4678377.570776255434541691
```

Where ursula2 is indeed receiving 2x rewards per period than ursula1 - so after the first period immediately after the increase, the problem resolves itself.

@vzotova hoping you can take a look at this when you are able to.

**Comment by derekpierre** [comment]
Greatly improved the suite of reward ratio tests including increase of stake via https://github.com/derekpierre/nucypher/commit/d6f15f2794e952ead2875acc26a8ecc2c3dcef7d. Now it is a standalone python module and tests for ratios are all lumped into one. The standalone test can be run via `pytest -s tests/contracts/integration/test_staking_reward_ratios.py`
. - ignore prior commit link in issue description.

Now the test more clearly shows that the issues are only during the `mint`, the first period immediately after the increase, values for all other periods are correct. Problematic assertions are only in the block that handles the case for the first period after the increase. These assertions can be skipped via setting `skipProblematicAssertionsAfterIncrease` on L82 to `True`. If set to `True` the test will fail the first period after the increase, if set to `False` the test won't fail but the staking reward values will be printed that show the ratio is incorrect for the period after the increase, and the assertions for all other periods hold.

```
>>> Increase ursula4 NU in period 10
...
>>> ursula1 reward calculated during period 11 for 10: 1403513.271232876630362507 NU
>>> ursula2 reward calculated during period 11 for 10: 1403513.271232876630362507 NU
>>> ursula3 reward calculated during period 11 for 10: 2807026.542465753260725014 NU
>>> ursula4 reward calculated during period 11 for 10: 686756.635616438315181253 NU <---- this is the problematic value
>>> ursula1 total rewards when increase occurred 15438645.983561642933987577 NU
>>> ursula1 total rewards when increase occurred 15438645.983561642933987577 NU
>>> ursula1 total rewards when increase occurred 30877291.967123285867975154 NU
>>> ursula1 total rewards when increase occurred 14721889.347945204618806323 NU
...
>>> ursula1 reward calculated during period 12 for 11: 1169594.392694063858635422 NU
>>> ursula2 reward calculated during period 12 for 11: 1169594.392694063858635422 NU
>>> ursula3 reward calculated during period 12 for 11: 2339188.785388127717270845 NU
>>> ursula4 reward calculated during period 12 for 11: 2339188.785388127717270845 NU
>>> ursula1 rewards since increase: 1169594.392694063858635422 NU
>>> ursula2 rewards since increase: 1169594.392694063858635422 NU
>>> ursula3 rewards since increase: 2339188.785388127717270845 NU
>>> ursula4 rewards since increase: 2339188.785388127717270845 NU
...
>>> ursula1 reward calculated during period 13 for 12: 1169594.392694063858635422 NU
>>> ursula2 reward calculated during period 13 for 12: 1169594.392694063858635422 NU
>>> ursula3 reward calculated during period 13 for 12: 2339188.785388127717270845 NU
>>> ursula4 reward calculated during period 13 for 12: 2339188.785388127717270845 NU
>>> ursula1 rewards since increase: 2339188.785388127717270844 NU
>>> ursula2 rewards since increase: 2339188.785388127717270844 NU
>>> ursula3 rewards since increase: 4678377.57077625543454169 NU
>>> ursula4 rewards since increase: 4678377.57077625543454169 NU
...
```

**Comment by derekpierre** [comment]
I can confirm that the same thing also happens when trying to increase a sub-stake via creating a new sub-stake, and then merging. Added another test in #2693 to reproduce the problem. The following is the output of that test:

```
>>> Increase ursula4 NU via merge in period 10
...
>>> ursula1 reward calculated during period 11 for 10: 1403513.271232876630362507 NU
>>> ursula2 reward calculated during period 11 for 10: 1403513.271232876630362507 NU
>>> ursula3 reward calculated during period 11 for 10: 2807026.542465753260725014 NU
>>> ursula4 reward calculated during period 11 for 10: 686756.635616438315181253 NU <---- this is the problematic value
>>> ursula1 total rewards when increase occurred 15438645.983561642933987577 NU
>>> ursula2 total rewards when increase occurred 15438645.983561642933987577 NU
>>> ursula3 total rewards when increase occurred 30877291.967123285867975154 NU
>>> ursula4 total rewards when increase occurred 14721889.347945204618806323 NU
...
>>> ursula1 reward calculated during period 12 for 11: 1169594.392694063858635422 NU
>>> ursula2 reward calculated during period 12 for 11: 1169594.392694063858635422 NU
>>> ursula3 reward calculated during period 12 for 11: 2339188.785388127717270845 NU
>>> ursula4 reward calculated during period 12 for 11: 2339188.785388127717270845 NU
>>> ursula1 rewards since increase: 1169594.392694063858635422 NU
>>> ursula2 rewards since increase: 1169594.392694063858635422 NU
>>> ursula3 rewards since increase: 2339188.785388127717270845 NU
>>> ursula4 rewards since increase: 2339188.785388127717270845 NU
...
>>> ursula1 reward calculated during period 13 for 12: 1169594.392694063858635422 NU
>>> ursula2 reward calculated during period 13 for 12: 1169594.392694063858635422 NU
>>> ursula3 reward calculated during period 13 for 12: 2339188.785388127717270845 NU
>>> ursula4 reward calculated during period 13 for 12: 2339188.785388127717270845 NU
```

**Comment by derekpierre** [comment]
Added a new test, linking to it here - https://github.com/nucypher/nucypher/pull/2693#issuecomment-861634984.

**Comment by derekpierre** [comment]
Closing as not fixed - but can be worked around (see #2693 ).

TL;DR:
The following procedure is an effective workaround:
1. Create a sub-stake
2. After 2 period commitments have occurred, merge the sub0stake

After chatting with @vzotova we probably won't fix the underlying increase operation bug since there is a workaround, and the prospective staking changes for Keanu may make any change moot - disclaimers should be sufficient.

Disclaimer for CLI added in #2693.
@vepkenez we probably need to update the disclaimer for the Staking UI - related to https://github.com/nucypher/nuit/pull/137.
